### ISSUE TITLE
Redis as in-memory store

### ISSUE THREAD

**Issue by tunght** [Ylianst/MeshCentral#3780]
Hi Ylianst,

I've been working on with MeshCentral lately and really like the project. 
I'm working on setting up server peering however I see that MeshCentral is relying on change stream from mongodb to update each server in-memory states. It's not working for us as we are using Postgres. 
I'm planning to add redis support to MeshCentral so that I can use peering with any database type and I see you had Redis in the code before so I'd like to check with you the reason why you removed redis? and would you condider to use Redis as the in-memory database?

Thanks
Tung

**Comment by Ylianst** [comment]
I am fully ok making use of Redis, I am not too familiar with it and I can learn. One thing however is that I want to make sure keep MeshCentral light on dependencies for people just starting out. So, as long as it does not add a hard dependency, I am 100% ok with it.

**Comment by archef2000** [comment]
I would love to use meshcentral with postgresql  using the peering feature. Is there any future where it will cone again?

**Comment by archef2000** [comment]
@si458 Is there any intrest in a PR for peering support via redis?

**Comment by si458** [comment]
I think we would have to support redis as a standalone database first before we support peering in redis.

However! If u are happy to spend the time making the changes and submitting a PR for the support of redis, then we aren't stopping you!

We would indeed give u a shout-out in the release notes for the support, tho hehe!

**Comment by archef2000** [comment]
What is send over ṕeering in MongoDB?

**Comment by si458** [comment]
@archef2000 I'm not 100% sure on the server peering,  from my understanding to server peering u need mongodb, the reason for this is because mongodb has replication build in when setup in a certain way, this allows ur database to fallover in the event of anything, also the data in ur database can be accessed from all your server too
Now in thinking about it, I'm not sure how redis would help? All it's data runs in memory! If u stop it u lose everything? So using it as a database wouldn't work
U would need to explain a little more of what u are thinking/planning for us to understand

**Comment by archef2000** [comment]
I thought that mongodb is used as a message broker between MC instances when enabeling peering. What does MC stop from being HA with databases like psotgres when there is a proxy infront?

**Comment by si458** [comment]
@archef2000 no meshcentral talks to other meshcentral servers over websockets,
the database is just a database. it store the information
https://ylianst.github.io/MeshCentral/meshcentral/#server-peering

**Comment by archef2000** [comment]
Sorry I misunderstood peering. Then why is MongoDB required for server peering?

**Comment by si458** [comment]
@archef2000 im not 100% sure im afraid, all i know is its require for server peering

**Comment by archef2000** [comment]
Yes, but I already have setup postgres ha and prefer it over MongoDB

**Comment by si458** [comment]
well unfortunately to use server peering it required mongodb
if you wanted a different database you need to open a new enhancement request in issues
and we will look into it but it wont be anytime soon as we have too much on at the moment

**Comment by tunght** [comment]
Mesh agents and browsers connect to 1 server, so when there are multiple servers, when a browser or a mesh agent connect to a server, it needs to notify other servers about the new connection. Meshcentral use mongodb change stream to  do that. Other database doesn't have that feature. In oder to make this feature agnostic to database, we just need to use redis to store all connections information like a "routing table" so that each server know which other server to forward data to. We don't need to make Redis as a database for other stuff, just for the "routing table"

**Comment by archef2000** [comment]
Yes that was my intend to use redis as notification broker/where each agent is connected to so it works with every database. (except a local file)

**Comment by tunght** [comment]
Also not only meshcentral store connections in memory, but it also load data of all devices data from database to store in memory whether or not the devices are online. So if someone modify the database ( like adding/ removing a new device ) the database will send a change notification to other servers to update their in memory state as well. Redis would help improve that as well

**Comment by archef2000** [comment]
@si458 Is there any work on that would love to see this feature.

**Comment by si458** [comment]
@archef2000 i started it but hit SO MANY road blocks because it requires a COMPLETE sourcecode rewrite
so its going to take a long time, for example
```
var node = obj.meshagents['node//xxxxxxxxxxxxxxxxxxxxxxxxxx'];
```
We cant access the object directly with redis
we have to query a function to get the value otherwise it uses variables which are then being stored in memory
so the is no point using redis at that point
(if that makes sense?)

**Comment by archef2000** [comment]
Are you working on that rewrite or is something else blocking it? If there is anything I can help with let me know. Thank you for your work.

**Comment by si458** [comment]
@archef2000 unfortunately its on the back burner at the moment as i have other bug fixes/features to do
i have been able to migrate a few variables from memory to redis!
but the is just so much code which does as i showed above
and redis doesnt work like that as it needs to be async and not sync in javascript land 👎

**Comment by archef2000** [comment]
Ok. Let me know if there are any updates.

**Comment by archef2000** [comment]
Still relevant
